File: app/src/main/java/org/calyxos/bellis/PostProvisioningHelper.java
/*
 * Copyright (C) 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.calyxos.bellis;

import android.app.admin.DevicePolicyManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.SharedPreferences;

import androidx.annotation.NonNull;

class PostProvisioningHelper {

    private static final String PREFS = "post-provisioning";
    private static final String PREF_DONE = "done";

    private static final String[] REQUIRED_PACKAGES = new String[]{
            "com.google.android.gms",
            "com.google.android.gsf",
            "com.android.vending",
            "org.fitchfamily.android.dejavu",
            "org.microg.nlp.backend.ichnaea",
            "org.microg.nlp.backend.nominatim",
            "com.stevesoltys.seedvault",
            "org.fdroid.fdroid",
            "org.chromium.chrome",
    };

    private final Context mContext;
    private final DevicePolicyManager mDevicePolicyManager;
    private final SharedPreferences mSharedPrefs;

    PostProvisioningHelper(@NonNull Context context) {
        mContext = context;
        mDevicePolicyManager =
                (DevicePolicyManager) context.getSystemService(Context.DEVICE_POLICY_SERVICE);
        mSharedPrefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE);
    }

    public void completeProvisioning() {
        if (isDone()) {
            return;
        }
        ComponentName componentName = BasicDeviceAdminReceiver.getComponentName(mContext);
        // This is the name for the newly created managed profile.
        mDevicePolicyManager.setProfileName(
                componentName,
                mContext.getString(R.string.app_name)
        );
        // We enable the profile here.
        mDevicePolicyManager.setProfileEnabled(componentName);

        for (String pkg : REQUIRED_PACKAGES) {
            mDevicePolicyManager.enableSystemApp(componentName, pkg);
        }

        mDevicePolicyManager.setBackupServiceEnabled(componentName, true);
    }

    public boolean isDone() {
        return mSharedPrefs.getBoolean(PREF_DONE, false);
    }
}

File: app/src/main/java/org/calyxos/bellis/PostProvisioningHelper.kt
/*
 * SPDX-FileCopyrightText: 2014 The Android Open Source Project
 * SPDX-FileCopyrightText: 2022 The Calyx Institute
 * SPDX-License-Identifier: Apache-2.0
 */

package org.calyxos.bellis

import android.app.admin.DevicePolicyManager
import android.content.Context
import android.content.Intent
import android.util.Log
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.OutOfQuotaPolicy
import androidx.work.WorkManager
import androidx.work.Worker
import androidx.work.WorkerParameters
import java.util.concurrent.atomic.AtomicBoolean

object PostProvisioningHelper {

    private val TAG = PostProvisioningHelper::class.java.simpleName
    private const val PREFS = "post-provisioning"
    private const val PREF_DONE = "done"
    private var postProvisioningInProgress = AtomicBoolean()

    // Default apps to enable on creation of new managed profile
    private val systemApps = listOf(
        "com.android.gallery3d",
        "com.android.vending",
        "com.android.camera2",
        "com.google.android.gms",
        "com.google.android.gsf",
        "com.stevesoltys.seedvault",
        "org.calyxos.eleven",
        "org.chromium.chrome",
        "org.fdroid.fdroid",
        "org.fitchfamily.android.dejavu",
        "org.microg.nlp.backend.ichnaea",
        "org.microg.nlp.backend.nominatim"
    )

    class SystemAppWorker(private val context: Context, workerParams: WorkerParameters) :
        Worker(context, workerParams) {
        override fun doWork(): Result {
            val componentName = BasicDeviceAdminReceiver.getComponentName(context)
            context.getSystemService(DevicePolicyManager::class.java)?.apply {
                systemApps.forEach {
                    try {
                        enableSystemApp(componentName, it)
                    } catch (e: IllegalArgumentException) {
                        Log.e(TAG, "Failed to enable $it")
                    }
                }
            }
            return Result.success()
        }
    }

    fun completeProvisioning(context: Context) {
        val isInProgress = postProvisioningInProgress.getAndSet(true)
        if (isInProgress) {
            return
        }
        if (!provisioningComplete(context)) {
            val devicePolicyManager = context.getSystemService(DevicePolicyManager::class.java)
            val componentName = BasicDeviceAdminReceiver.getComponentName(context)
            devicePolicyManager.apply {
                setProfileName(componentName, context.getString(R.string.app_name))

                WorkManager.getInstance(context).enqueue(
                    OneTimeWorkRequestBuilder<SystemAppWorker>().setExpedited(
                        OutOfQuotaPolicy.RUN_AS_NON_EXPEDITED_WORK_REQUEST
                    ).build()
                )
            }

            val sharedPreferences = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            sharedPreferences.edit().putBoolean(PREF_DONE, true).apply()

            devicePolicyManager.setProfileEnabled(componentName)

            launchSUW(context)
        }
        postProvisioningInProgress.set(false)
    }

    private fun provisioningComplete(context: Context): Boolean {
        val sharedPreferences = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        return sharedPreferences.getBoolean(PREF_DONE, false)
    }

    private fun launchSUW(context: Context) {
        val setupWizard = "org.lineageos.setupwizard"
        val setupWizardActivity = ".SetupWizardActivity"

        val intent = Intent(Intent.ACTION_MAIN).apply {
            setClassName(setupWizard, setupWizard + setupWizardActivity)
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        }
        context.startActivity(intent)
    }
}
